### ISSUE TITLE
Feature request for debugging `eval` commands (phpArgumentInterpreter.php)

### ISSUE THREAD

**Issue by RieksJ** [AmpersandTarski/Ampersand#622]
After a long time of debugging, I found the following message in FireBug:

![image](https://cloud.githubusercontent.com/assets/8522589/22181164/07f2507a-e086-11e6-953e-c79935526f0a.png)

Apparently, I am using a php-function that is not properly evaluated. 
It would be nice if the logs were to contain some clues about what is actually happening here.

Can we have this `eval` function (in `phpArgumentInterpreter.php`) log the text that it is trying to evaluate in the debug log?

**Comment by Michiel-s** [comment]
Hi @RieksJ,

I understand the issue and why it took you a while to find the issue. I advise you to add 'checking the response of an API call' as one of the first steps in your debugging flow. 

The point here is that parse errors (in this case due to dynamic code evaluation) cannot be handled with a user defined error handler (see: http://php.net/manual/en/function.set-error-handler.php).

Furthermore, 

> The eval() language construct is very dangerous because it allows execution of arbitrary PHP code. Its use thus is discouraged. If you have carefully verified that there is no other option than to use this construct, pay special attention not to pass any user provided data into it without properly validating it beforehand.
see: http://php.net/manual/en/function.eval.php

Therefore, I propose that we rewrite the exec engine functionality regarding the phpArgumentInterpreter.

### call_user_func:
Instead of using `...;{php}code;...` in your exec engine violation text, which calls `eval(code)`, we will allow function calls only. You should specify `...;{php}phpfunctionname(arg,arg);...` which calls a (user)defined function and if applicable returns the needed value. This allows you to create and use your own functions, just like the exec engine function itself.

We'll check for the existence of the function and throw an error if it does not exists.

Of course you still have to provide valid syntax for those functions, but you can check those files design time with `php -l filename.php`.

### example:
we often use `{php}date(DATE_ISO8601)` in an exec engine rule to get the current date. This code doesn't have to change, because now it will call the function date(), instead of running eval(date()). The result is the same, but more secure.

**Comment by RieksJ** [comment]
That's good. Note however that I used the following VIOLATION:
~~~
VIOLATION (TXT "{EX}_;InsPair_;ttInstance_;TText_;", TGT I, TXT "_;TTPhrase_;"
              , TXT "{php}str_replace('['.'", SRC ttName  -- text to replace (all occurences will be replaced)
                             , TXT "'.']','", SRC ttValue -- text by which it is to be replace 
                             , TXT "','", TGT ttInstance  -- string in which the replacements should take place
                             , TXT "')" 
          ,TXT "{EX} InsPair;ttValueIsUsedInTTexts;TText;", SRC I, TXT ";TText;", TGT I
          ,TXT "{EX} InsPair;ttUsedValue;TText;", SRC I, TXT ";TTValue;", SRC ttValue
          )
~~~
which actually DOES call a function, namely `str_replace`, which is similar to calling the function `date()`.
So basically, I would not have to change anything, would I?

**Comment by Michiel-s** [comment]
In the refactored backend, parse errors are catched and a proper error message is shown. This will solve the issue of not easily seeing where the error is. Furthermore I propose to not change anything else (for now).
